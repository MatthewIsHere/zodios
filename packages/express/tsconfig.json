{
  // ts config for es 2021
  "compilerOptions": {
    "target": "ES2020",
    "module": "CommonJS",
    "moduleResolution": "Node",
    "outDir": "./lib",
    "alwaysStrict": true,
    "strict": true,
    "esModuleInterop": true,
    "sourceMap": true,
    "declaration": true
  },
  "exclude": ["node_modules", "lib", "../../examples"]
}
