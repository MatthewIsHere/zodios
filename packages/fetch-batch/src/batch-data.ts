type BatchDataForeachCallback = (
  request: Request,
  contentId: string,
  batchdata: BatchData
) => void;

export class BatchData {
  #requests = new Map<string, Request>();
  #contentIdSuffix = `${Date.now()}@zodios.org`;
  #boundary = `batch__${Date.now()}__batch`;
  #encoder = new TextEncoder();

  constructor() {}

  get boundary() {
    return this.#boundary;
  }

  addRequest(request: Request) {
    const contentId = `request-${this.#requests.size + 1}-${
      this.#contentIdSuffix
    }`;
    this.#requests.set(contentId, request);
    return contentId;
  }

  getHeaders(init?: HeadersInit) {
    const headers = new Headers(init);
    headers.set("Content-Type", `multipart/mixed; boundary=${this.#boundary}`);
    return headers;
  }

  [Symbol.iterator]() {
    return this.#requests.entries();
  }

  hasRequest(contentId: string) {
    return this.#requests.has(contentId);
  }

  requests() {
    return this.#requests.values();
  }

  contentIds() {
    return this.#requests.keys();
  }

  entries() {
    return this.#requests.entries();
  }

  getRequest(contentId: string) {
    return this.#requests.get(contentId);
  }

  forEach(callback: BatchDataForeachCallback) {
    this.#requests.forEach((request, contentId) =>
      callback(request, contentId, this)
    );
  }

  find(request: Request) {
    for (const [contentId, req] of this.#requests.entries()) {
      if (req === request) {
        return contentId;
      }
    }
  }

  stream() {
    const iterator = this.#encodedIterator();
    return new ReadableStream<Uint8Array>({
      async pull(controller) {
        const { value, done } = await iterator.next();
        if (done) {
          controller.close();
        } else {
          controller.enqueue(value);
        }
      },
    });
  }

  #formatHeaders(headers: Headers) {
    let result = "";
    for (const [name, value] of headers) {
      result += `${name}: ${value}\r\n`;
    }
    result += "\r\n";
    return result;
  }

  *#encodePart(contentId: string) {
    const headers = new Headers();
    headers.set("Content-ID", `<${contentId}>`);
    headers.set("Content-Type", "application/http; msgtype=request");
    yield this.#encoder.encode(this.#formatHeaders(headers));
  }

  async *#encodeRequest(request: Request) {
    const url = new URL(request.url);
    yield this.#encoder.encode(
      `${request.method} ${url.pathname}${url.search}${url.hash} HTTP/1.1\r\n`
    );
    yield this.#encoder.encode(this.#formatHeaders(request.headers));
    if (request.body) {
      const reader = request.body.getReader();
      while (true) {
        const { done, value } = await reader.read();
        if (done) {
          break;
        }
        yield value;
      }
    }
  }

  async *#encodedIterator() {
    const boundary = this.#encoder.encode(`--${this.#boundary}\r\n`);
    const boundaryClose = this.#encoder.encode(`--${this.#boundary}--\r\n`);

    for (const [contentId, request] of this.#requests) {
      yield boundary;
      yield* this.#encodePart(contentId);
      yield* this.#encodeRequest(request);
      yield this.#encoder.encode("\r\n");
    }
    yield boundaryClose;
  }
}
