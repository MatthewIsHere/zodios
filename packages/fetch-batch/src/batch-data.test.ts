import express from "express";
import type { AddressInfo } from "net";
import { BatchData } from "./batch-data";

function getUser1Req(port: number) {
  return new Request(`http://localhost:${port}/get`, {
    method: "GET",
    headers: new Headers({
      Accept: "application/json; charset=utf-8",
    }),
  });
}

function renameUser2Req(port: number) {
  return new Request(`http://localhost:${port}/patch`, {
    method: "PATCH",
    headers: new Headers({
      Accept: "application/json; charset=utf-8",
      "Content-Type": "application/json; charset=utf-8",
    }),
    body: JSON.stringify({
      name: "John Doe",
    }),
  });
}

describe("BatchData", () => {
  let app: express.Express;
  let server: ReturnType<typeof app.listen>;
  let port: number;

  beforeAll(() => {
    const app = express();
    app.use(express.raw({ type: "multipart/mixed" }));

    app.post("/batch", (req, res) => {
      console.log("received batch request");
      res.json({
        url: req.url,
        method: req.method,
        headers: req.headers,
        data: req.body.toString(),
      });
    });

    server = app.listen(0);
    port = (server.address() as AddressInfo).port;

    jest.useFakeTimers();
    jest.setSystemTime(new Date(2023, 1, 1));
  });

  afterAll((done) => {
    jest.useRealTimers();
    server.close(done);
  });

  it("should be able to batch requests", async () => {
    const getUser1 = getUser1Req(port);
    const renameUser2 = renameUser2Req(port);

    const body = new BatchData();
    body.addRequest(getUser1);
    body.addRequest(renameUser2);

    const batched = await fetch(`http://localhost:${port}/batch`, {
      method: "POST",
      headers: body.getHeaders(),
      body: body.stream(),
    }).then((response) => response.json());

    expect(batched.url).toBe("/batch");
    expect(batched.method).toBe("POST");
    expect(batched.headers["content-type"]).toContain("multipart/mixed");
    expect(batched.data).toMatchSnapshot();
  });

  it("should be able to iterate over requests", async () => {
    const getUser1 = getUser1Req(port);
    const renameUser2 = renameUser2Req(port);

    const body = new BatchData();
    body.addRequest(getUser1);
    body.addRequest(renameUser2);

    const requests: Request[] = [];
    for (const [, request] of body) {
      requests.push(request);
    }
    expect(requests).toEqual([getUser1, renameUser2]);

    const requests2: Request[] = [];
    for (const [, request] of body.entries()) {
      requests2.push(request);
    }
    expect(requests2).toEqual([getUser1, renameUser2]);

    expect(Array.from(body.requests())).toEqual([getUser1, renameUser2]);

    expect(Array.from(body.requestIds())).toEqual([
      "request-1-1675206000000@zodios.org",
      "request-2-1675206000000@zodios.org",
    ]);

    body.forEach((request, contentId) => {
      expect(request).toBe(body.getRequest(contentId));
    });

    expect(body.getRequestId(getUser1)).toBe(
      "request-1-1675206000000@zodios.org"
    );
    expect(body.getRequestId(renameUser2)).toBe(
      "request-2-1675206000000@zodios.org"
    );

    expect(body.boundary).toBe("batch__1675206000000__batch");
  });

  it("should be able to get request by content id", async () => {
    const getUser1 = getUser1Req(port);
    const renameUser2 = renameUser2Req(port);

    const body = new BatchData();
    body.addRequest(getUser1);
    body.addRequest(renameUser2);

    expect(body.getRequest("response-request-1-1675206000000@zodios.org")).toBe(
      getUser1
    );
    expect(body.getRequest("response-request-2-1675206000000@zodios.org")).toBe(
      renameUser2
    );
  });
});
