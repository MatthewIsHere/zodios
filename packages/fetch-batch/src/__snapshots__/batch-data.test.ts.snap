// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchData makeBatchData should be able to batch requests 1`] = `
"--batch__1675206000000__batch
content-id: <request-1-1675206000000@zodios.org>
content-type: application/http; msgtype=request

GET /get HTTP/1.1
accept: application/json; charset=utf-8


--batch__1675206000000__batch
content-id: <request-2-1675206000000@zodios.org>
content-type: application/http; msgtype=request

PATCH /patch HTTP/1.1
accept: application/json; charset=utf-8
content-type: application/json; charset=utf-8

{"name":"John Doe"}
--batch__1675206000000__batch--
"
`;

exports[`BatchData makeReactNativeBatchData should be able to batch requests 1`] = `
"--batch__1675206000000__batch
content-id: <request-1-1675206000000@zodios.org>
content-type: application/http; msgtype=request

GET /get HTTP/1.1
accept: application/json; charset=utf-8


--batch__1675206000000__batch
content-id: <request-2-1675206000000@zodios.org>
content-type: application/http; msgtype=request

PATCH /patch HTTP/1.1
accept: application/json; charset=utf-8
content-type: application/json; charset=utf-8

{"name":"John Doe"}
--batch__1675206000000__batch--
"
`;
